Add ElementCenteredSubdomainData for Schwarz solver

nilsvu opened the pull request:
## Proposed changes

An element-centered subdomain consists of a central element and overlap
regions with all neighboring elements. This class holds data on such a
subdomain. It supports vector space operations (addition and scalar
multiplication) and an inner product, which allow to use this data type with
linear solvers (e.g. `LinearSolver::Serial::Gmres`).

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Add ElementCenteredSubdomainData for Schwarz solver

An element-centered subdomain consists of a central element and overlap
regions with all neighboring elements. This class holds data on such a
subdomain. It supports vector space operations (addition and scalar
multiplication) and an inner product, which allow to use this data type with
linear solvers (see e.g. `LinearSolver::Serial::Gmres`).

kidder commented:
@geoffrey4444 @moxcodes can you review this PR?  Please assign yourself if you can

moxcodes commented on src/ParallelAlgorithms/LinearSolver/Schwarz/ElementCenteredSubdomainData.hpp:
```diff
@@ -0,0 +1,214 @@
+ * An element-centered subdomain consists of a central element and overlap
+ * regions with all neighboring elements. This class holds data on such a
+ * subdomain. It supports vector space operations (addition and scalar
+ * multiplication) and an inner product, which allow to use this data type with
```
`which allow to use this` -> `which allows the use of this`

moxcodes commented on src/ParallelAlgorithms/LinearSolver/Schwarz/ElementCenteredSubdomainData.hpp:
```diff
@@ -0,0 +1,214 @@
+struct ElementCenteredSubdomainData {
+  static constexpr size_t volume_dim = Dim;
+  using ElementData = Variables<TagsList>;
+  using OverlapData = ElementData;
```
For clarification, is the reason this couldn't be accomplished with a single `Variables` because the overlap data will in general be a different size than the element data?

nilsvu replied:
To apply a DG operator to the subdomain data I need to pass `Variables` to our routines that apply differentiation matrices, compute fluxes and so on, and these `Variables` have to contain data on a single LGL mesh. That's why I store the `Variables` for the central element and for the overlaps with each neighbor separately. It would be nice to store all data contiguously and get "views" into this data that I can pass along as `Variables`, but I don't think that's currently possible. If I will find that this subdomain data class is relevant for performance I may have to implement something like that though.

moxcodes commented on src/ParallelAlgorithms/LinearSolver/Schwarz/ElementCenteredSubdomainData.hpp:
```diff
@@ -0,0 +1,214 @@
+};
+
+template <size_t Dim, typename LhsTagsList, typename RhsTagsList>
+decltype(auto) operator+(
```
I think you only want to define these operators for lhs and rhs with the same taglists, right?

nilsvu replied:
The `Variables` class supports arithmetics with tag lists that differ by prefixes, so I'm just carrying over that functionality here. It can be useful when working with prefixed tags.

moxcodes commented on src/ParallelAlgorithms/LinearSolver/Schwarz/ElementCenteredSubdomainData.hpp:
```diff
@@ -0,0 +1,214 @@
+      LinearSolver::Schwarz::ElementCenteredSubdomainData<Dim, TagsListOut>;
+  static SPECTRE_ALWAYS_INLINE SubdomainDataOut
+  apply(const SubdomainDataIn& input, const double value) noexcept {
+    SubdomainDataOut output{input.element_data.number_of_grid_points()};
```
can you default construct this instead of passing in the number of grid points? I'm concerned this allocates twice when it could just allocate once on the next line.

moxcodes commented on tests/Unit/ParallelAlgorithms/LinearSolver/Schwarz/Test_ElementCenteredSubdomainData.cpp:
```diff
@@ -0,0 +1,89 @@
+    return subdomain_data;
+  };
+  auto subdomain_data1 = make_subdomain_data({1., 2., 3.}, {4., 5.}, {6.});
+  auto subdomain_data2 = make_subdomain_data({2., 1., 0.}, {1., 2.}, {3.});
```
I believe as currently written `subdomain_data2` can be const also.

moxcodes reviewed (commented):
Looks pretty good -- only a few small comments.

moxcodes reviewed (commented):
Looks good -- go ahead and squash.  Looks like CI is reporting a few clang-tidy complaints that you can squash along with the fixup.

nilsvu commented:
Thanks for your review @moxcodes, I squashed and fixed the clang-tidy complaints. @kidder could you perhaps do the second review in place of @geoffrey4444?

geoffrey4444 reviewed (commented):
My apologies for not replying sooner! I can review later this week if no one else can, but if someone else can take this one, please go ahead

kidder commented on src/ParallelAlgorithms/LinearSolver/Schwarz/ElementCenteredSubdomainData.hpp:
```diff
@@ -0,0 +1,214 @@
+  using OverlapData = ElementData;
+
+  ElementCenteredSubdomainData() = default;
+  ElementCenteredSubdomainData(const ElementCenteredSubdomainData&) = default;
```
noexcept these two

kidder commented on src/ParallelAlgorithms/LinearSolver/Schwarz/ElementCenteredSubdomainData.hpp:
```diff
@@ -0,0 +1,214 @@
+  ElementCenteredSubdomainData& operator+=(
+      const ElementCenteredSubdomainData<Dim, RhsTagsList>& rhs) noexcept {
+    element_data += rhs.element_data;
+    for (auto& id_and_overlap_data : overlap_data) {
```
use a structured bindings  e.g. `auto& [id, data] : overlap_data` for these loops

kidder commented on src/ParallelAlgorithms/LinearSolver/Schwarz/ElementCenteredSubdomainData.hpp:
```diff
@@ -0,0 +1,214 @@
+  return os;
+}
+
+template <size_t Dim, typename LhsTagsList, typename RhsTagsList>
```
I don't think you should allow different TagsList for comparisons

kidder reviewed (changes requested):
you can squash the changes

nilsvu commented:
Implemented your changes @kidder, rebased and squashed. Could you take another look? Thanks for reviewing @moxcodes and @kidder!
